lumen/ai/llm.py

```diff
@@ -169,40 +181,83 @@ class OpenAI(Llm):
     organization = param.String()
 
     _models = {
-        'gpt-4': {
-            'model_name': 'gpt-4'
-        },
-        'gpt-3.5-turbo': {
-            'model_name': 'gpt-3.5-turbo'
-        }
+        "gpt-4": {"model_name": "gpt-4"},
+        "gpt-3.5-turbo": {"model_name": "gpt-3.5-turbo"},
     }
 
     @property
     def _client_kwargs(self):
-        return {'model': self.model_name}
+        return {"model": self.model_name}
 
     def _init_model(self):
         import openai
+
         model_kwargs = {}
         if self.base_url:
-            model_kwargs['base_url'] = self.base_url
+            model_kwargs["base_url"] = self.base_url
         if self.api_key:
-            model_kwargs['api_key'] = self.api_key
+            model_kwargs["api_key"] = self.api_key
         if self.organization:
-            model_kwargs['organization'] = self.organization
+            model_kwargs["organization"] = self.organization
         self._model = openai.OpenAI(**model_kwargs)
         self._raw_client = self._model.chat.completions.create
         self._client = self._create_client(self._raw_client)
 
 
+class OutlinesLlama(Llama):
+
+    mode = param.Selector(default=Mode.JSON)
+
+    def _create_client(self, create):
+        return create
+
+    def _init_model(self):
+        from outlines import models
+        from huggingface_hub import hf_hub_download
+        from llama_cpp import Llama
+        from llama_cpp.llama_speculative import LlamaPromptLookupDecoding
+
+        draft_model = LlamaPromptLookupDecoding(num_pred_tokens=10)
+        llama_kwargs = dict(
+            model_path=hf_hub_download(self.repo, self.model_file),
+            n_gpu_layers=-1,
+            n_ctx=8192,
+            chat_format=self.chat_format,
+            logits_all=False,
+            verbose=False,
+        )
+        llama = Llama(**llama_kwargs)
+        # can't use llama object https://github.com/outlines-dev/outlines/issues/730
+        self._model = pn.state.as_cached(
+            "Llama", partial(models.LlamaCpp, draft_model=draft_model), **llama_kwargs
+        )
+        self._raw_client = llama.create_chat_completion_openai_v1
+        self._client = self._create_client(self._model)
+
+    def _generate_json(
+        self,
+        create: "models.LlamaCpp",
+        messages: list | str,
+        system: str = "",
+        **kwargs,
+    ):
+        from outlines import generate
+
+        generator = generate.json(create, kwargs.pop("response_model"))
+        message = "\n\n".join(
+            [f"System: {system}"]
+            + [f"{message['role']}: {message['content']}" for message in messages],
+        )
```

Comment:
It doesn't accept the OpenAI format; I don't think this is how I should be formatting it.

Reply:
I think I should be applying something like https://huggingface.co/blog/chat-templates but I need to load transformers too...